efi/fw_load_handler.go

```diff
@@ -20,9 +20,251 @@
+	}
+
+	ctx.MeasureVariable(secureBootPCR, name.GUID, name.Name, db)
+	return db, nil
+}
+
+func (h *fwLoadHandler) measureAuthorizedSignatureDb(ctx pcrBranchContext) error {
+	data, err := h.measureSignatureDb(ctx, Db)
+	if err != nil {
+		return err
+	}
+
+	db, err := efi.ReadSignatureDatabase(bytes.NewReader(data))
+	if err != nil {
+		return xerrors.Errorf("cannot decode signatures: %w", err)
+	}
+	ctx.FwContext().Db = &secureBootDB{Name: Db, Contents: db}
+	return nil
+}
+
+func (h *fwLoadHandler) measureSecureBootPolicyPreOS(ctx pcrBranchContext) error {
+	ctx.ResetPCR(secureBootPCR)
+
+	// This hard-codes a profile that will only work on devices with secure boot enabled,
+	// deployed mode on (where UEFI >= 2.5), without a UEFI debugger enabled and which
+	// measure events in the correct order.
+	ctx.MeasureVariable(secureBootPCR, efi.GlobalVariable, sbStateName, []byte{1})
+	if _, err := h.measureSignatureDb(ctx, PK); err != nil {
+		return xerrors.Errorf("cannot measure PK: %w", err)
+	}
+	if _, err := h.measureSignatureDb(ctx, KEK); err != nil {
+		return xerrors.Errorf("cannot measure KEK: %w", err)
+	}
+	if err := h.measureAuthorizedSignatureDb(ctx); err != nil {
+		return xerrors.Errorf("cannot measure db: %w", err)
+	}
+	if _, err := h.measureSignatureDb(ctx, Dbx); err != nil {
+		return xerrors.Errorf("cannot measure dbx: %w", err)
+	}
+	// TODO: Support optional dbt/dbr databases
+
+	// Retain any verification events associated with pre-OS components such as UEFI drivers
+	// or system preparation applications. Note that these make a profile inherently fragile.
+	// See the comment in measureBootManagerCodePreOS regarding event ordering.
+	events := h.log.Events
+	foundOsPresent := false
+	foundSecureBootSeparator := false
+
+	for len(events) > 0 {
+		e := events[0]
+		events = events[1:]
+
+		switch {
+		case e.PCRIndex < secureBootPCR && e.EventType == tcglog.EventTypeSeparator:
+			// pre-OS to OS-present signal
+			foundOsPresent = true
+		case e.PCRIndex == secureBootPCR && e.EventType == tcglog.EventTypeSeparator:
+			// end of secure boot configuration signal
+			if foundSecureBootSeparator {
+				return errors.New("unexpected separator")
+			}
+			ctx.ExtendPCR(secureBootPCR, tpm2.Digest(e.Digests[ctx.PCRAlg()]))
+			foundSecureBootSeparator = true
+		case e.PCRIndex == secureBootPCR && e.EventType == tcglog.EventTypeEFIVariableAuthority:
+			// secure boot verification event
+			if !foundSecureBootSeparator {
+				return errors.New("unexpected verification event")
+			}
+			digest := tpm2.Digest(e.Digests[ctx.PCRAlg()])
+			ctx.FwContext().AppendVerificationEvent(digest)
+			ctx.ExtendPCR(secureBootPCR, digest)
+		case e.PCRIndex == secureBootPCR && e.EventType == tcglog.EventTypeEFIVariableDriverConfig:
+			// ignore: part of the secure boot configuration
+		case e.PCRIndex == secureBootPCR:
+			return fmt.Errorf("unexpected event type (%v) found in log", e.EventType)
```

Comment:
should the confusion=>error checks here be tested?

Reply:
I've added some extra tests here.